File: Algorithms/leetcode/Ex_0019_Remove_Nth_Node_From_End_of_List/Soln_2_Two_Pointers.cpp
#include <iostream>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution_2 {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *p = head;
        ListNode *q = head;
        // 先让 p 向前走 n 步（题目说n <= sz，所以不必考虑n大于链表长度的问题，即while条件不需要考虑p=nullptr的情况）
        while (n--) {
            p = p->next;
        }
        // 此时 p 指向了第 k+1 个节点（这里解释为什么不让 p 指向第 n 个，因为当 p 指向第 n+1 个节点，此时 q 指向第 1 个节点，这时 p 与 q 之间的距离恰好是n，即 q 要往前走 n 步才能能追上 p。那么当 p 向前推进到空节点时，q 还需要再往前走 n 步才能为空，即 q 与 p 之间还有 n-1 个节点，所以此时 q 指向的的是倒数第 n 个节点）（一定配合题解参考中的插图看！）
        // 举例：
        // 1 2 3 4 5 6 7 8 9 null     (设 n = 5)
        // p
        // q
        // 
        // 让 p 先走了 n 步:
        // 1 2 3 4 5 6 7 8 9 null
        // q         p
        // 
        // 依次往前推进
        // 1 2 3 4 5 6 7 8 9 null
        //   q ^ ^ ^   p ^ ^  ^
        //     q | |     p |  |
        //       q |       p  |
        //         q          p
        // 注意现在 q 和 p之间还有n-1 个节点，q还需要往前走n步才能追上p
        //
        // 让 p 和 q 同时向前推进，并记录 q 前一个节点
        ListNode *prev_q = nullptr;
        while (p != nullptr) {
            p = p->next;
            prev_q = q;
            q = q->next;
        }
        // 删除节点 q
        if (q == head) {
            // 删除的是头节点
            return head->next;
        } else {
            prev_q->next = q->next;
        }
        return head;
    }
};

// 参考题解（强烈推荐）：https://leetcode-cn.com/problems/linked-list-cycle/solution/yi-wen-gao-ding-chang-jian-de-lian-biao-wen-ti-h-2/

File: Algorithms/leetcode/Ex_0160_Intersection_of_Two_Linked_Lists/Soln_2_Two_Pointers.java
package Algorithms.leetcode.Ex_0160_Intersection_of_Two_Linked_Lists;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
class Solution_2 {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        ListNode pA = headA;
        ListNode pB = headB;
        while (pA != pB) {
            pA = (pA == null) ? headB : pA.next;
            pB = (pB == null) ? headA : pB.next;
        }
        // 如果两条链表有相交点，那么此时 pA 和 pB 都位于相交的起始节点
        // 如果两条链没有相交点，则 pA / pB 会同时指向 headB / headA 链表的末尾节点后的 null，也会退出 while 循环
        return pA;
    }
}

// 参考题解（推荐）：https://leetcode-cn.com/problems/intersection-of-two-linked-lists/solution/tu-jie-xiang-jiao-lian-biao-by-user7208t/

